### PR TITLE
[#85] Claims can access resource function

### LINKED ISSUE #85: [FEATURE] Claims can access resource function
**Is your feature request related to a problem? Please describe.**
Currently, there is no consistent method which determines whether a user's token claims are sufficient to access resources with an `appID`, `orgID`, and `system` status flag.

**Describe the solution you'd like**
Add a function to `tokenauth.Claims` which determines whether the set of claims grants access to a resource having the three parameters mentioned above. 

**Describe alternatives you've considered**
Each service using the auth library could implement their own functions to do this, but this will likely result in inconsistencies across the system and unnecessary code duplication.

### PR DESCRIPTION
**Pull request by roberlander2**
## Description
This PR adds a `CanAccess` function to `tokenauth.Claims`, which determines whether the claims grant access to a resource with a given `appID`, `orgID`, and `system` status.

**Resolves #85**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [x] Within a week
- [ ] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [x] I have added tests related to my changes.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

**Review by edthedev — DISMISSED** [review]
Appreciate the additional test coverage here!

File: tokenauth/token.go [hunk 1]
```diff
@@ -72,6 +72,30 @@ func (c Claims) Scopes() []authorization.Scope {
...
+// CanAccess returns an error if the claims do not grant access to a resource with the given appID, orgId, and system status
+func (c Claims) CanAccess(appID string, orgID string, system bool) error {
+	var err error
+	// if system admin or not a system resource, err = nil
+	if !c.System && system {
+		return errors.New("non-system admin access is forbidden")
+	}
+
+	// only system admins may access resources applying to all apps, all orgs
+	if appID == authutils.AllApps && orgID == authutils.AllOrgs {
```
Comment (shurwit) [inline tokenauth/token.go:84]: I think this will still cause an issue in the case where `system = falase`

File: tokenauth/token.go [hunk 2]
```diff
@@ -72,6 +72,30 @@ func (c Claims) Scopes() []authorization.Scope {
...
+		return err
```
Comment (shurwit) [inline tokenauth/token.go:85]: Won't `err` always be nil?

File: tokenauth/token_test.go [hunk 3]
```diff
@@ -77,6 +77,43 @@ func getSampleExpiredClaims() *tokenauth.Claims {
...
+		args    args
+		wantErr bool
+	}{
+		{"success on system access all_all", args{&systemClaims, authutils.AllApps, authutils.AllOrgs, true}, false},
+		{"success on system access all_org", args{&systemClaims, authutils.AllApps, "org1", false}, false},
+		{"success on system access app_all", args{&systemClaims, "app1", authutils.AllOrgs, true}, false},
+		{"success on system access app_org", args{&systemClaims, "app1", "org1", true}, false},
+		{"error on access other app_org", args{&systemClaims, "app1", "org2", true}, true},
+
+		{"error on admin access all_all", args{&adminClaims, authutils.AllApps, authutils.AllOrgs, true}, true},
```
Comment (shurwit) [inline tokenauth/token_test.go:101]: Shouldn't this test case have `system = false`?

**Review by shurwit — COMMENTED** [review]
Thanks @roberlander2, I think this looks close but I have a couple questions about the logic below

**Review by shurwit — APPROVED** [review]
Thanks @roberlander2, LGTM!

**Pull request merged**